src/gridsynth.rs

```diff
@@ -406,61 +522,101 @@ fn search_for_solution(
     let (epsilon_region, unit_disk, transformed) = setup_regions_and_transform(
         config.theta.clone(),
         config.epsilon.clone(),
         config.verbose,
         config.measure_time,
+        phase,
     );
 
-    search_for_solution(&epsilon_region, &unit_disk, &transformed, config)
+    search_for_solution(&epsilon_region, &unit_disk, &transformed, config, phase)
 }
 
 pub fn gridsynth_gates(config: &mut GridSynthConfig) -> GridSynthResult {
-    let start_total = if config.measure_time {
-        Some(Instant::now())
-    } else {
-        None
-    };
-
-    let u_approx = gridsynth(config);
-
-    let start_decompose = if config.measure_time {
-        Some(Instant::now())
-    } else {
-        None
-    };
-    let gates = decompose_domega_unitary(u_approx);
+    // let start_total = if config.measure_time {
+    //     Some(Instant::now())
+    // } else {
+    //     None
+    // };
+
+    // let start_decompose = if config.measure_time {
+    //     Some(Instant::now())
+    // } else {
+    //     None
+    // };
+
+    if !config.up_to_phase {
+        // exact synthesis only
+        let u_approx = gridsynth(config, PhaseMode::Exact);
+        let gates = decompose_domega_unitary(u_approx);
+
+        // Perform validation check, if required.
+        let (error, is_correct) = match config.compute_error {
+            true => {
+                let (e, is_ok) =
+                    compute_error(&gates, &config.theta, &config.epsilon, PhaseMode::Exact);
+                (Some(e), Some(is_ok))
+            }
+            false => (None, None),
+        };
 
-    if let Some(start) = start_decompose {
-        if config.measure_time {
-            debug!(
-                "time of decompose_domega_unitary: {:.3} ms",
-                start.elapsed().as_secs_f64() * 1000.0
-            );
-        }
-    }
-    if let Some(start) = start_total {
-        if config.measure_time {
-            debug!(
-                "total time: {:.3} ms",
-                start.elapsed().as_secs_f64() * 1000.0
-            );
+        GridSynthResult {
+            gates,
+            global_phase: false,
+            error,
+            is_correct,
         }
-    }
-
-    // Peform validation check, if required.
-    let (error, is_correct) = if config.compute_error {
-        let (error, is_correct) = compute_error(&gates, &config.theta, &config.epsilon);
-        (Some(error), Some(is_correct))
     } else {
-        (None, None)
-    };
-
-    GridSynthResult {
-        gates,
-        error,
-        is_correct,
+        // exact synthesis
+        let u_approx = gridsynth(config, PhaseMode::Exact);
+        let gates_exact = decompose_domega_unitary(u_approx);
+
+        // also shifted synthesis
+        let u_approx = gridsynth(config, PhaseMode::Shifted);
+        let gates_shifted = decompose_domega_unitary(u_approx);
+
+        if gates_exact.len() <= gates_shifted.len() {
```

Comment:
The algorithm is supposed to return the solution with lowest T count. But this returns the one with the lowest gate count.

Reply:
Fixed in [d933ff7](https://github.com/qiskit-community/rsgridsynth/pull/31/commits/d933ff7071121afd0511f9e2bac2c2476a368028).